BUG: Partial fix for docstring validation for parameters

ChiefMilesEdgeworth opened the pull request:
This PR addresses #20298, does not close.
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

This changes the method used to find arguments from `inspect.getfullargspec` to `inspect.signature` and changes the handling accordingly. It also now checks for "*args, **kwargs" and other variations as a doc parameter, and splits them up so they can be checked against.

ChiefMilesEdgeworth committed:
BUG: Use signature instead of getfullargspec 

Using signature allows the validation to get to the root function when a function is decorated. This reduced a significant amount of the errors in the string.py file

ChiefMilesEdgeworth committed:
Use signature instead of getfullargspec 

Using signature allows the validation to get to the root function when a function is decorated. This reduced a significant amount of the errors in the string.py file

ChiefMilesEdgeworth committed:
BUG: Account for args and kwargs on same line

Check for variations on "*args, **kwargs" while loading in the documentation parameters, and load them in separately if found.

ChiefMilesEdgeworth committed:
Account for args and kwargs on same line

Check for variations on "*args, **kwargs" while loading in the documentation parameters, and load them in separately if found.

ChiefMilesEdgeworth committed:
Styling Changes

Pep8 conformity

ChiefMilesEdgeworth committed:
Styling Changes

Pep8 conformity

ChiefMilesEdgeworth committed:
DOC: Idxmin/max fixed 

Those functions cause issues during the testing. This has most likely been covered by some PR merged already.

ChiefMilesEdgeworth committed:
Merge remote-tracking branch 'origin/fix_docstring_validation' into fix_docstring_validation

WillAyd commented:
There are tests for the doctoring validator - if making changes there also will need a test case(s)

Nabel0721 commented:
Thank you! I'm working out the bugs introduced right now.

ChiefMilesEdgeworth committed:
BUG: Fixed desc being loaded as list instead of str

Desc was being loaded in without being turned into a string. This was causing issues when indexes were used to check capitalization and the final period.

Nabel0721 commented:
Would you like me to write tests for decorated functions? That was the main point addressed by this PR. If so, just let me know, and I can do that before this is merged.

ChiefMilesEdgeworth committed:
CLN: Missed styling changes

For some reason black didn't catch the validation script the first time. It should be fixed now.

Nabel0721 commented:
Never mind, I found it. Sorry for all the really little commits, I just recently started working with pandas, so I'm getting used to all the small things that need to be done first.

ChiefMilesEdgeworth committed:
TST: Added tests for decorated functions

Tests added to test_validate_docstrings that check that it handles decorated functions as predicted.

pep8speaks commented:
Hello @Nabel0721! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`scripts/tests/test_validate_docstrings.py`](https://github.com/pandas-dev/pandas/blob/5a13341b5c09fa4a2e8725f95929d311bbe4eec3/scripts/tests/test_validate_docstrings.py):

> [Line 25:74](https://github.com/pandas-dev/pandas/blob/5a13341b5c09fa4a2e8725f95929d311bbe4eec3/scripts/tests/test_validate_docstrings.py#L25): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 26:73](https://github.com/pandas-dev/pandas/blob/5a13341b5c09fa4a2e8725f95929d311bbe4eec3/scripts/tests/test_validate_docstrings.py#L26): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

ChiefMilesEdgeworth committed:
CLN: Missed whitespace changes

Fixed trailing whitespace on two lines.

Nabel0721 commented:
I added tests for decorators. Look over it and see if there are any more you'd like me to add. If I have an idea of what to make I should be able to do it no problem, I'm just not sure what exactly would need to be added. The changes I made to the validation script were just to account for oversights, not to add functionality, so there weren't any new features to test.

datapythonista reviewed (commented):
Thanks for working on this.

There are two unrelated changes here, one changing how we obtain the paramters from the docstring, and one from the signature. Do you mind addressing them in separate PRs? The one of the docstrings is quite simple I think, but I don't quite understand the other one. And in case we change our mind and we want to revert, we won't want to revert both.

Regarding `inspect.signature`, can you point me to an example where this change is fixing a validation error. I'm sure you know what you're doing, but would like to understand what's going on, and I fail to see if we're fixing bugs, or just using a nicer API.

Thanks!

Nabel0721 commented:
So, here's a code example to show what the difference is between `getfullargspec` and `signature`.
```python3
from functools import wraps
from inspect import getfullargspec, signature


def decorator(func):
    @wraps(func)
    def wrapper(*args, **kwargs):
        """
        This is hidden documentation that we don't want
        """
        print("The wrapper is doing stuff!")
        return func(*args, **kwargs)
    return wrapper

@decorator
def add(a, b, *, absolute=False):
    """
    This method adds a and b, with the absolute value taken
    """
    return abs(a+b) if absolute else a+b

    
print(f"Using getfullargspec: {getfullargspec(add)}")
print(f"Using signature {signature(add)}")
```

The following script prints out:
```
Using getfullargspec: FullArgSpec(args=[], varargs='args', varkw='kwargs', defaults=None, kwonlyargs=[], kwonlydefaults=None, annotations={})
Using signature (a, b, *, absolute=False)
```

`getfullargspec` only finds the signature of the decorator, since it doesn't search any deeper. However, `signature` checks if the object has the `__wrapped__` attribute (which is assigned in the by the @wraps decorator to the decorated function), and if it does it checks the signature of that object instead. Using `signature` allows us to properly get the signature of decorated functions, which is important since we have decorators everywhere. For instance, there is a decorator for deprecating a specific kwarg. That was causing trouble before, because it overrode the signature of the function below it. This would solve that issue.

Nabel0721 commented:
I'll put together a separate PR for the splitting on ", ". You're right that these two issues are pretty distinctly separate, and should be kept that way.

datapythonista commented:
Thanks a lot for the detail explanation on signature vs getfullargsspec. Is there any case in pandas where this is an issue right now, and that it's fixed with this PR?

We've got `_to_original_callable` which I think it's taking care of this. I guess it could be simplified with this change, but it'd be nice to know how this change is affecting the reported docstrings.

Nabel0721 commented:
Most of the PR02 errors we're getting from `pandas.Series.str` are the result of this. There are other places too, I remember finding them as I was looking around at PR02 errors. They usually pop up where the PR02 error says variables aren't defined that are in the signature of the function. If it's decorated, then it gets *args, **kwargs instead of whatever is supposed to be there.

We could run the check on the object returned by `to_original_callable`, but if we do that we should just remove that function and use `inspect.unwrap` instead. It does the same thing but covers more cases than we currently do. I was thinking of submitting that idea later anyways, since it cuts out a whole function and would result in better functionality (catching more cases).

datapythonista commented:
Ah, I see. Do you think it makes sense to add one of the Series.str methods to the unit tests? Also, it'd probably make sense to generate the errors json with and without these changes, and find the differences, and see if the result is the expected one.

Nabel0721 commented:
Yeah, I did that when I first made the changes, but I can send two files with the differences.

Nabel0721 commented:
One file is taken from current master, so there'll be some changes that were fixed by other PRs. The main ones to look at are under Series.str.
[decorator_fix.txt](https://github.com/pandas-dev/pandas/files/3695475/decorator_fix.txt)
[master.txt](https://github.com/pandas-dev/pandas/files/3695476/master.txt)

A lot of the errors are gone after applying the fix.

ChiefMilesEdgeworth committed:
TST: Revert change to arg handling

Reverting back to master's version so the issue can be addressed in another PR.

datapythonista commented:
Sounds great. Let's split the two different things, and merge them then.

Nabel0721 commented:
I'm probably not going to have the other one ready till tomorrow, it ran into a small issue on the testing and I need to head to bed. If this one passes the tests it should be good to go though.

ChiefMilesEdgeworth committed:
Merge pull request #1 from pandas-dev/master

Update Local Version

datapythonista commented on scripts/tests/test_validate_docstrings.py:
```diff
@@ -53,6 +87,21 @@ def sample(self):
         return random.random()
 
+    @Decorators.good_decorator
+    def decorated_sample(self):
```
shouldn't we have parameters here, to see if this test is really working?

datapythonista commented on scripts/validate_docstrings.py:
```diff
@@ -436,16 +436,20 @@ def signature_parameters(self):
+            elif data.kind == inspect.Parameter.VAR_KEYWORD:
+                params.append("**" + parameter)
+            else:
+                params.append(parameter)
```
I think this would be more readable if we implement a function like:
```python
def with_stars(name, param_info):
    if param_info.kind == inspect.Parameter.VAR_POSITIONAL:
        return '*{}'.format(name)
    ...
    return name

params = [with_stars(name, param_info) for name, param_info in sign.parameters.items())]
```

datapythonista reviewed (commented):
looks good, just couple of last comments

ChiefMilesEdgeworth committed:
Better test and readability improvement

I added a parameter to the test so that it's more visible, and changed the validation script to use a generator straight into a tuple. I also switched off of using `items`, since it's a set and therefore might give different orders for the parameters.

ChiefMilesEdgeworth committed:
Merge remote-tracking branch 'origin/master'

ChiefMilesEdgeworth committed:
Fixed PEP8 Requirements

Fixed line length and whitespace requirements

ChiefMilesEdgeworth committed:
Merge remote-tracking branch 'upstream/master'

ChiefMilesEdgeworth committed:
Merge remote-tracking branch 'upstream/master'

Updating to most recent version

ChiefMilesEdgeworth committed:
Merge branch 'master' into fix_docstring_validation

Update branch

datapythonista commented on scripts/tests/test_validate_docstrings.py:
```diff
@@ -664,6 +715,19 @@ def list_incorrect_parameter_type(self, kind):
         pass
 
+    @Decorators.missing_wraps
+    def bad_decorator(self, kind: bool):
```
Sorry I didn't think about this in the previous reviews, but this feels a bit wrong. We're testing that we're unable to get the correct signature for this.

Since I guess we can't know whether the documented parameters are wrong, or the function is incorrectly wrapped, I think probably the best is simply to remove this test. If this test fails would be great news, so I don't think we want to have it. Or we can have it as an `xfail`.

@WillAyd what do you think?

Nabel0721 replied:
Imo xfail would probably be best. This test demonstrates the problem that was solved, and it passing means that the script should be worked on again to remove extraneous code.

datapythonista replied:
I may be missing something, but if we had the function `bad_decorator` in pandas, what we should document in Parameters is `kind`, right? Since our validation will fail if `kind` is documented, I'd say that we solved other problems, but this test is something that ideally we'd like to fix.

Am I wrong?

Nabel0721 replied:
I think you're right, I'm just not sure how I'd go about fixing it. Which error would we like to produce for the test?

datapythonista replied:
I'd probably just delete the test. Or as I said, I'd move it to the passing tests, and then xfail it. But probably not worth.

Nabel0721 replied:
Alright, I'll remove it and push another commit.

ChiefMilesEdgeworth committed:
Remove bad parameter test

Remove a test that doesn't quite examine the error we'd like.

datapythonista commented on scripts/tests/test_validate_docstrings.py:
```diff
@@ -68,6 +87,23 @@ def sample(self):
         """
         return random.random()
 
+    @Decorators.good_decorator
```
Can you use `@functools.lru_cache` instead, so we don't need to implement a custom decorator. I guess `lru_cache` is well wrapped.

When importing, I think it's a better practice to `import functools` than `from functools import wraps, lru_cache`.

datapythonista reviewed (commented):
just one last comment, to make things simpler, and we're good to go

ChiefMilesEdgeworth committed:
Use lru_cache instead of custom decorator

Switched to using a builtin decorator

datapythonista reviewed (approved):
lgtm, thanks @Nabel0721, nice improvement

WillAyd commented:
Read through history but not entirely clear - what does this end up improving upon?

Nabel0721 commented:
Whenever we use a decorator on a function that follows normal decorator conventions (wrapping the original function, passing in the arguments and returning the result with some added functionality) we were losing the signature of the original function. The result was that the script would look for *args, **kwargs or whatever the signature was of the wrapper instead of the actual function's signature, and that would be used in the comparison to the docstring parameters. This resolves that by using a more up-to-date function from the builtin library `inspect`. It checks if a function has the `__wrapped__` attribute, and if so calls itself recursively on the function in that attribute. So, with this change we'll see most of the errors due to decorators go away, although a few of the decorators need a bit of work since they don't use the @wraps decorator on their wrapper function (which sets the __wrapped__ attribute along with a few other things). If you need me to go into more specifics on any part of this, just let me know. At this point I've got a really good grasp on how decorators work and what the `inspect.signature` does.

WillAyd commented:
Makes sense. I think generally we are inconsistent in how we maintain signatures, annotations, etc... with wrapped functions, so would welcome general ideas you have for that

datapythonista commented:
We were doing it manually until now with `_to_original_callable`. @Nabel0721 I guess you'll see if it can be fully removed now in a follow up, right?

Nabel0721 commented:
Yeah, honestly we could just replace the entire function with a call to
inspect.unwrap

On Mon, Oct 14, 2019 at 12:07 PM Marc Garcia <notifications@github.com>
wrote:

> We were doing it manually until now with _to_original_callable. @Nabel0721
> <https://github.com/Nabel0721> I guess you'll see if it can be fully
> removed now in a follow up, right?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/28765?email_source=notifications&email_token=ACT76UTUNBLA2NKH3XHJRELQOSKLHA5CNFSM4I46DLWKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBFLYEA#issuecomment-541768720>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACT76UQBXJPUMQONNHHNRGLQOSKLHANCNFSM4I46DLWA>
> .
>
-- 
Sincerely,

*Nathan Abel*
Kettering University - Computer Science
989-780-3050

datapythonista commented:
Great, please tag me when you open the PR. Thanks!

Nabel0721 commented:
Ideally, we should only be working with signatures and wrapped functions using the builtin tools from the inspect library whenever possible. I'm fairly certain that this is what numpy uses as well, so it will create consistency with the imported docstring stuff from them.

datapythonista commented:
I didn't see that Python had that unwapping functionality, that's why I implemented `_to_original_callable`. If we can get the actual function just calling the Python sdlib, that would be great, less code to maintain.
